### ISSUE TITLE
Coordinates passed as sets raise with a bad error message

### ISSUE THREAD

**Issue by max-sixty** [pydata/xarray#1691]
If a coordinate is passed as a `set`, xr raises with a bad error message:

```python

In [12]: xr.Dataset(dict(date=[1,2,3], sec={4}))
---------------------------------------------------------------------------
MissingDimensionsError                    Traceback (most recent call last)
<ipython-input-12-40ccdd94e21f> in <module>()
----> 1 xr.Dataset(dict(date=[1,2,3], sec={4}))

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/dataset.py in __init__(self, data_vars, coords, attrs, compat)
    360             coords = {}
    361         if data_vars is not None or coords is not None:
--> 362             self._set_init_vars_and_dims(data_vars, coords, compat)
    363         if attrs is not None:
    364             self.attrs = attrs

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/dataset.py in _set_init_vars_and_dims(self, data_vars, coords, compat)
    378
    379         variables, coord_names, dims = merge_data_and_coords(
--> 380             data_vars, coords, compat=compat)
    381
    382         self._variables = variables

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/merge.py in merge_data_and_coords(data, coords, compat, join)
    363     objs = [data, coords]
    364     explicit_coords = coords.keys()
--> 365     return merge_core(objs, compat, join, explicit_coords=explicit_coords)
    366
    367

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/merge.py in merge_core(objs, compat, join, priority_arg, explicit_coords, indexes)
    425     coerced = coerce_pandas_values(objs)
    426     aligned = deep_align(coerced, join=join, copy=False, indexes=indexes)
--> 427     expanded = expand_variable_dicts(aligned)
    428
    429     coord_names, noncoord_names = determine_coords(coerced)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/merge.py in expand_variable_dicts(list_of_variable_dicts)
    211                     var_dicts.append(coords)
    212
--> 213                 var = as_variable(var, name=name)
    214                 sanitized_vars[name] = var
    215

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/xarray/core/variable.py in as_variable(obj, name)
    103                 'cannot set variable %r with %r-dimensional data '
    104                 'without explicit dimension names. Pass a tuple of '
--> 105                 '(dims, data) instead.' % (name, data.ndim))
    106         obj = Variable(name, obj, fastpath=True)
    107     else:

MissingDimensionsError: cannot set variable 'sec' with 0-dimensional data without explicit dimension names. Pass a tuple of (dims, data) instead.
```
But OK if a list:
```python
In [13]: xr.Dataset(dict(date=[1,2,3], sec=[4]))
Out[13]:
<xarray.Dataset>
Dimensions:  (date: 3, sec: 1)
Coordinates:
  * date     (date) int64 1 2 3
  * sec      (sec) int64 4
Data variables:
    *empty*


```
#### Problem description

There may be reasons to not allow sets: they're not ordered, so unless you're constructing your data using the `coords`, you'll get random results

The error message should be better though. And I would vote to handle sets the same as lists

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 17.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.0rc1-2-gf83361c
pandas: 0.21.0
numpy: 1.13.3
scipy: 0.19.1
netCDF4: None
h5netcdf: None
Nio: None
bottleneck: 1.2.1
cyordereddict: None
dask: None
matplotlib: 2.0.2
cartopy: None
seaborn: 0.8.1
setuptools: 36.5.0
pip: 9.0.1
conda: None
pytest: 3.2.3
IPython: 6.2.1
sphinx: None
</details>

**Comment by shoyer** [comment]
We convert lists into array, but it's not immediately clear to me what to do with set objects. For example, NumPy treats them as scalars: 
```
In [17]: np.array({1})
Out[17]: array({1}, dtype=object)
```

**Comment by max-sixty** [comment]
Closed by #2331
